### PR TITLE
feat(fs)!: add --only-link option and changed --only-file and --only-d…

### LINKED ISSUE #516: feat: include symlinks in `--only-files`
...or something smarter. At least let's keep this in mind, could be a footgun for users.
---
For file and directory counts

```
eza --only-dirs | wc -l
eza --only-files | wc -l
```

Apparently symlinks are not included as files, so \#directory entries – \#files – \#dirs = \#symlinks?

_Originally posted by @holmesrichards in https://github.com/eza-community/eza/issues/341#issuecomment-1757724823_

### LINKED ISSUE #517: feat: add `--only-links`
> yup and maybe we could do `--only-links` at this point 😄

_Originally posted by @MartinFillon in https://github.com/eza-community/eza/issues/516#issuecomment-1758337469_

### PR DESCRIPTION
**Pull request by MartinFillon**
…ir behaviour

--only-link now only shows links

BREAKING CHANGE:
--only-file now only shows files and links to files --only-dir now only shows directories and links to directories

closes #517
closes #516

### CODE REVIEW

**Review by cafkafk — CHANGES_REQUESTED** [review]
You need the PR summaries to be "<something>!: <something>" (notice the `!` for breaking change).

**Comment by MartinFillon** [comment]
my bad

File: src/options/flags.rs [hunk 1]
```diff
@@ -35,6 +35,8 @@ pub static GIT_IGNORE:  Arg = Arg { short: None, long: "git-ignore",           t
 pub static DIRS_FIRST:  Arg = Arg { short: None, long: "group-directories-first",  takes_value: TakesValue::Forbidden };
 pub static ONLY_DIRS:   Arg = Arg { short: Some(b'D'), long: "only-dirs", takes_value: TakesValue::Forbidden };
 pub static ONLY_FILES:   Arg = Arg { short: Some(b'f'), long: "only-files", takes_value: TakesValue::Forbidden };
+pub static ONLY_LINKS:   Arg = Arg { short: None, long: "only-links", takes_value: TakesValue::Forbidden };
+pub static NO_LINKS:   Arg = Arg { short: None, long: "no-links", takes_value: TakesValue::Forbidden };
```
Comment (cafkafk) [inline src/options/flags.rs:39]: This part should be aligned like the lines above (could you also realign line 37?)

**Review by cafkafk — CHANGES_REQUESTED** [review]
This looks good code wise, just needs one nit and completions for zsh, bash, fish, and nushell

**Review by cafkafk — CHANGES_REQUESTED** [review]
You need to add the `!` to the commit summary of the breaking change as well

**Review by cafkafk — CHANGES_REQUESTED** [review]
Maybe I'm doing something wrong, but this says the arguments doesn't exist (`eza: Unknown argument --only-links`)? Maybe the git history got messed up :woman_shrugging: ? Also this probably needs completions.

**Comment by MartinFillon** [comment]
That is clearly not supposed to happen gonna investigate

**Comment by MartinFillon** [comment]
Ok in fact I was dumb, but that is also due to the parser being so great

**Comment by MartinFillon** [comment]
Okay I can't seem to find the way to run the command `TRYCMD=overwrite cargo test` in nix could you help me with that @cafkafk ?

**Comment by cafkafk** [comment]
> Okay I can't seem to find the way to run the command `TRYCMD=overwrite cargo test` in nix could you help me with that @cafkafk ?

You can't currently do that, but we have `just idump` that should update the tests

**Comment by MartinFillon** [comment]
got this error `eza> touch: cannot touch 'tests/itest/dirlink': No such file or directory`

**Comment by MartinFillon** [comment]
@cafkafk could you run `just idump` for me

**Comment by cafkafk** [comment]
> got this error `eza> touch: cannot touch 'tests/itest/dirlink': No such file or directory`

Is this link supposed to be broken? If not, I think the problem here is that the structure inside of the nix builder (that runs the tests) is different. This is also why we need to incorporate dynamic test directory generations through the just file. I suppose then that testing is blocking this, but we might just not include those tests for now as long as we ensure they work manually

**Comment by MartinFillon** [comment]
Okay gonna delete them to be sure

File: README.md [hunk 2]
```diff
@@ -324,8 +324,10 @@ eza’s options are almost, but not quite, entirely unlike `ls`’s.
 - **-r**, **--reverse**: reverse the sort order
 - **-s**, **--sort=(field)**: which field to sort by
 - **--group-directories-first**: list directories before other files
-- **-D**, **--only-dirs**: list only directories
-- **-f**, **--only-files**: list only files
+- **-D**, **--only-dirs**: list only directories and links to directories
+- **-f**, **--only-files**: list only files and links to files
+- **--only-links**: list only symbolic links
```
Comment (cafkafk) [inline README.md:329]: Trying this branch I get `eza: Unknown argument --only-links`

**Review by cafkafk — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (MartinFillon) [reply README.md:329]: gonna investigate this

Comment (MartinFillon) [reply README.md:329]: This might be due to merge errors but fixed it

**Review by cafkafk — CHANGES_REQUESTED** [review]
This looks like it works (and is quite useful), just needs completions as far as I can tell.

**Comment by MartinFillon** [comment]
Here are the completions.

**Review by cafkafk — APPROVED** [review]
Integration test pass, further:

```
ln -s Justfile justfilefile
cargo run -- -l --only-dirs
cargo run -- -l --only-files
cargo run -- -l --only-dirs --no-links
cargo run -- -l --only-files --no-links
cargo run -- -l --no-links
```
All work.

But 
```
cargo run -- -l --only-dirs --only-links
cargo run -- -l --only-files --only-links
```
Don't work as I expected. Perhaps we should have them display only dirs that are links, and only files that are links.

...sorry for the long review process, this should be the last thing before this get merged!

**Review by cafkafk — CHANGES_REQUESTED** [review]
Sorry, that should have been requesting changes.

**Comment by MartinFillon** [comment]
> ...sorry for the long review process, this should be the last thing before this get merged!

It is okay we need to be sure before merging anything, long review helps with that.
> Don't work as I expected. Perhaps we should have them display only dirs that are links, and only files that are links.

Didn't thought of this use case gonna implement it

**Review by cafkafk — CHANGES_REQUESTED** [review]
When running 
```
cargo run -- -l --no-links
```
I still get links now.

Also, for me at least, neither of:
```
cargo run -- -l --only-dirs --only-links
cargo run -- -l --only-files --only-links
```
work by showing only links.

Also, integration tests seem like they broke somewhere

**Comment by MartinFillon** [comment]
Yeah it tests should break as it is a breaking change, I am gonna take a look tho

**Comment by MartinFillon** [comment]
Are you sure you are on the last version @cafkafk
20231204_23h04m21s_grim
20231204_23h04m38s_grim
your right for `eza --no-links` tho

**Comment by MartinFillon** [comment]
fixed the last one lasting

**Review by PThorpe92 — APPROVED** [review]
Code looks good, briefly tested and it works :+1:

**Review by cafkafk — CHANGES_REQUESTED** [review]
Just checked on latest version, all work except:

```bash
cargo run -- -l --only-dirs --only-links
cargo run -- -l --only-files --only-links
```

They also show dirs and files that aren't links

**Comment by PThorpe92** [comment]
> Just checked on latest version, all work except:
> 
> ```shell
> cargo run -- -l --only-dirs --only-links
> cargo run -- -l --only-files --only-links
> ```
> 
> They also show dirs and files that aren't links

:sweat_smile: lol good catch I didn't test them together

**Comment by MartinFillon** [comment]
made some changes @cafkafk if you wanna see

**Review by cafkafk — CHANGES_REQUESTED** [review]
There seems to be some conflicts, after fix I'll take a look :+1:

**Comment by MartinFillon** [comment]
Gonna fix all of that when I have time

**Comment by MartinFillon** [comment]
Superseeded by #1028

**Pull request closed without merging**